frame/support/src/storage/counted_map.rs

```diff
@@ -0,0 +1,314 @@
+use codec::{FullCodec, Decode, EncodeLike, Encode};
+use crate::{
+	storage::{
+		StorageMap, StorageValue,
+		StorageAppend, StorageTryAppend, StorageDecodeLength, StoragePrefixedMap,
+		generator::{StorageMap as StorageMapT, StorageValue as StorageValueT},
+	},
+};
+use sp_std::prelude::*;
+use sp_runtime::traits::Saturating;
+
+/// A wrapper around a `StorageMap` and a `StorageValue<u32>` to keep track of how many items are in
+/// a map, without needing to iterate all the values.
+///
+/// This storage item has additional storage read and write overhead when manipulating values
+/// compared to a regular storage map.
+///
+/// For functions where we only add or remove a value, a single storage read is needed to check if
+/// that value already exists. For mutate functions, two storage reads are used to check if the
+/// value existed before and after the mutation.
+///
+/// Whenever the counter needs to be updated, an additional read and write occurs to update that
+/// counter.
+pub struct CountedStorageMap<Key, Value, Map, Counter>(
+	core::marker::PhantomData<(Key, Value, Map, Counter)>
+);
+
+impl<Key, Value, Map, Counter>
+	StorageMapT<Key, Value>
+	for CountedStorageMap<Key, Value, Map, Counter>
```

Comment:
this is not correct, storage::generator::StorageMap implements storage::StorageMap,
Somebody can now call `<CountedStorageMap as storage::StorageMap>::remove()` and it will bypass the count update.

In framev2 we don't need to use generator anymore, the generator were some helpful trait for the decl_storage macro.
Now we should better implement StorageMap directly manually, unfortunately that is not possible as it conflict with the implementation by the generator for `T`.

So I'm not sure of the solution but we should get rid of generator conflict. Maybe modifying decl_storage so that it calls a macro which implements the traits StorageMap using the generator.